pkg/autodiscovery/listeners/snmp.go

```diff
@@ -0,0 +1,356 @@
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+
+	"github.com/soniah/gosnmp"
+)
+
+func init() {
+	Register("snmp", NewSNMPListener)
+}
+
+// SNMPListener implements SNMP discovery
+type SNMPListener struct {
+	sync.RWMutex
+	newService chan<- Service
+	delService chan<- Service
+	stop       chan bool
+	config     util.SNMPListenerConfig
+	services   map[string]Service
+	devices    map[string][]string
+}
+
+// SNMPService implements and store results from the Service interface for the SNMP listener
+type SNMPService struct {
+	adIdentifier string
+	entityID     string
+	deviceIP     string
+	creationTime integration.CreationTime
+}
+
+// Make sure SNMPService implements the Service interface
+var _ Service = &SNMPService{}
+
+type snmpSubnet struct {
+	adIdentifier  string
+	config        util.SNMPConfig
+	defaultParams *gosnmp.GoSNMP
+	startingIP    net.IP
+	currentIP     net.IP
+	network       net.IPNet
+	cacheKey      string
+}
+
+// NewSNMPListener creates a SNMPListener
+func NewSNMPListener() (ServiceListener, error) {
+	var snmpConfig util.SNMPListenerConfig
+	err := config.Datadog.UnmarshalKey("snmp_listener", &snmpConfig)
+	if err != nil {
+		return nil, err
+	}
+	return &SNMPListener{
+		services: map[string]Service{},
+		devices:  map[string][]string{},
+		stop:     make(chan bool),
+		config:   snmpConfig,
+	}, nil
+}
+
+// Listen periodically refreshes devices
+func (l *SNMPListener) Listen(newSvc chan<- Service, delSvc chan<- Service) {
+	// setup the I/O channels
+	l.newService = newSvc
+	l.delService = delSvc
+
+	go l.checkDevices()
+}
+
+func (l *SNMPListener) loadCache(config util.SNMPConfig, adIdentifier string, cacheKey string) {
+	cacheValue, err := persistentcache.Read(cacheKey)
+	if err != nil {
+		log.Errorf("Couldn't read cache for %s: %s", cacheKey, err)
+		return
+	}
+	if cacheValue == "" {
+		return
+	}
+	var devices []net.IP
+	if err = json.Unmarshal([]byte(cacheValue), &devices); err != nil {
+		log.Errorf("Couldn't unmarshal cache for %s: %s", cacheKey, err)
+		return
+	}
+	for _, deviceIP := range devices {
+		entityID := config.Digest(deviceIP.String())
+		l.createService(deviceIP.String(), adIdentifier, entityID)
+	}
+}
+
+func (l *SNMPListener) writeCache(cacheKey string, adIdentifier string) {
+	l.Lock()
+	defer l.Unlock()
+	devices := l.devices[adIdentifier]
+	cacheValue, err := json.Marshal(devices)
+	if err != nil {
+		log.Errorf("Couldn't marshal cache: %s", err)
+		return
+	}
+	if err = persistentcache.Write(cacheKey, string(cacheValue)); err != nil {
+		log.Errorf("Couldn't write cache: %s", err)
+	}
+}
+
+// Don't make it a method, to be overridden in tests
+var worker = func(l *SNMPListener, jobs <-chan snmpSubnet) {
+	for {
+		select {
+		case <-l.stop:
+			log.Debug("Stopping SNMP worker")
+			return
+		case subnet := <-jobs:
+			log.Debugf("Handling IP %s", subnet.currentIP.String())
+			l.checkDevice(subnet.adIdentifier, subnet.currentIP.String(), subnet.config, subnet.defaultParams)
```

Comment:
This could be a good place to add successful checks to a set on the `SNMPListener` object and if the `deviceIP` within the current job already exists, then skip the check.

Reply:
👍  The current `deleteService` / `deviceFailures` approach addresses this comment.